dace/transformation/passes/constant_propagation.py

```diff
@@ -69,40 +74,85 @@ def apply_pass(self, sdfg: SDFG, _, initial_symbols: Optional[Dict[str, Any]] =
+            self._collect_constants_for_region(sdfg, arrays, in_consts, pre_consts, post_consts, out_consts)
 
             # Keep track of replaced and ambiguous symbols
             symbols_replaced: Dict[str, Any] = {}
             remaining_unknowns: Set[str] = set()
 
             # Collect symbols from symbol-dependent data descriptors
             # If there can be multiple values over the SDFG, the symbols are not propagated
-            desc_symbols, multivalue_desc_symbols = self._find_desc_symbols(sdfg, per_state_constants)
+            desc_symbols, multivalue_desc_symbols = self._find_desc_symbols(sdfg, in_consts)
 
             # Replace constants per state
-            for state, mapping in optional_progressbar(per_state_constants.items(),
-                                                       'Propagating constants',
-                                                       n=len(per_state_constants),
+            for block, mapping in optional_progressbar(in_consts.items(), 'Propagating constants', n=len(in_consts),
                                                        progress=self.progress):
+                if block is sdfg:
+                    continue
+
                 remaining_unknowns.update(
                     {k
                      for k, v in mapping.items() if v is _UnknownValue or k in multivalue_desc_symbols})
                 mapping = {
                     k: v
                     for k, v in mapping.items() if v is not _UnknownValue and k not in multivalue_desc_symbols
                 }
-                if not mapping:
-                    continue
-
-                # Update replaced symbols for later replacements
-                symbols_replaced.update(mapping)
+                out_mapping = {
+                    k: v
+                    for k, v in out_consts[block].items() if v is not _UnknownValue and k not in multivalue_desc_symbols
+                }
 
-                # Replace in state contents
-                state.replace_dict(mapping)
-                # Replace in outgoing edges as well
-                for e in sdfg.out_edges(state):
-                    e.data.replace_dict(mapping, replace_keys=False)
+                if mapping:
+                    # Update replaced symbols for later replacements
+                    symbols_replaced.update(mapping)
+
+                    if isinstance(block, SDFGState):
+                        # Replace in state contents
+                        block.replace_dict(mapping)
+                    elif isinstance(block, AbstractControlFlowRegion):
+                        block.replace_dict(mapping, replace_in_graph=False, replace_keys=False)
+
+                if out_mapping:
+                    # Replace in outgoing edges as well
+                    for e in block.parent_graph.out_edges(block):
+                        e.data.replace_dict(out_mapping, replace_keys=False)
+                
+                if isinstance(block, LoopRegion):
```

Comment:
I don't like the special case here but I think it's necessary. Could this be refactored to a separate method like _propagate_loop_region?